pytorch_lightning/callbacks/base.py

```diff
@@ -29,6 +29,9 @@ class Callback(abc.ABC):
     Subclass this class and override any of the relevant hooks
     """
 
+    def on_model_parallel_setup(self, trainer, pl_module: LightningModule) -> None:
+        """Called before model parallel accelerator setup"""
+
```

Comment:
@SeanNaren n00b q: why is this needed as a callback hook too?

Reply:
Allows us to do this:

For FSDP:
https://github.com/PyTorchLightning/pytorch-lightning/blob/feat/fsdp/pytorch_lightning/plugins/training_type/fully_sharded.py#L128-L147

For DeepSpeed:

https://github.com/PyTorchLightning/pytorch-lightning/blob/feat/ds_update/pytorch_lightning/plugins/training_type/deepspeed.py#L284

So for those really large models that we'd like to shard instantly, this provides a context manager to instantiate your layers. In FSDP, you'll have to call auto_wrap/wrap yourself, in DeepSpeed this is done for you (each module is automatically sharded, in FSDP we have a policy in place).

Reason it's done like this is because we do not have a distributed environment set up till the `Trainer` is called. then upon fit, we can setup the environment! Let me know if you do have ideas though!

Reply:
But those two snippets you shared don't use `on_model_parallel_setup`, they use `model_parallel_context`.

I think the question is more along the lines of "when would you override `on_model_parallel_setup`?"

Reply:
The short answer is when you'd like to shard your model instantly (to save initialization time or load a very large model).

Here is a hard case for DeepSpeed: https://github.com/SeanNaren/minGPT/blob/stage3/mingpt/model.py#L155-L161

We'll eventually write docs once the integrations are merged to explain the use cases in greater details.

Reply:
wouldn't a name like `configure_sharded_model` or `configure_parallel_model` be more appropriate?